src/drivers/net/mcst/l_e1000.c

```diff
@@ -0,0 +1,728 @@
+	soft_reset &= ~(SRST);
+	e1000_write_mgio_csr(ep, soft_reset); /* stop software reset */
+
+	/* Setup HW (MAC), also known as "Physical" address.
+	 * because unknown now, assumed to be 10:20:30:40:50:60 HEX */
+	for (i = 0; i < 6; i++) {
+		dev->dev_addr[i] = l_base_mac_addr[i];
+	}
+
+	fdx = 1; mii = 1; gmii = 1;
+	ep->mii_if.full_duplex = fdx;
+	ep->mii_if.supports_gmii = gmii;
+	ep->mii_if.phy_id_mask = 0x1f;
+	ep->mii_if.reg_num_mask = 0x1f;
+	ep->mii_if.phy_id = 0x01;
+
+	/* Setup PHY MII/GMII enable */
+	val = e1000_mii_readreg(dev, ep->mii_if.phy_id, PHY_AUX_CTRL);
+	log_boot("PHY reg # 0x12 (AUX_CTRL) : after reset : 0x%x\n", val);
+	val &= ~(RGMII_EN_1 | RGMII_EN_0);
+	e1000_mii_writereg(dev, ep->mii_if.phy_id, PHY_AUX_CTRL, val);
+	/* Setup PHY 10/100/1000 Link on 10M Link */
+	val = e1000_mii_readreg(dev, ep->mii_if.phy_id, PHY_LED_CTRL);
+	log_boot("PHY reg # 0x13 (LED_CTRL) : after reset : 0x%x\n", val);
+	val |= RED_LEN_EN;
+	e1000_mii_writereg(dev, ep->mii_if.phy_id, PHY_LED_CTRL, val);
+	val = e1000_mii_readreg(dev, ep->mii_if.phy_id, PHY_LED_CTRL);
+	log_boot("PHY reg # 0x13 (LED_CTRL) : after led is : 0x%x\n", val);
+
+	val = e1000_mii_readreg(dev, ep->mii_if.phy_id, 0);
+	log_boot("PHY reg # 0x0  after reset : 0x%x\n", val);
+	val = e1000_mii_readreg(dev, ep->mii_if.phy_id, 0x1);
+	log_boot("PHY reg # 0x1  after reset : 0x%x\n", val);
+	val = e1000_mii_readreg(dev, ep->mii_if.phy_id, 0x1);
+	log_boot("PHY reg # 0x1  after reset : 0x%x\n", val);
+	val = e1000_mii_readreg(dev, ep->mii_if.phy_id, 0x10);
+	log_boot("PHY reg # 0x10  after reset : 0x%x\n", val);
+	val = e1000_mii_readreg(dev, ep->mii_if.phy_id, 0x11);
+	log_boot("PHY reg # 0x11  after reset : 0x%x\n", val);
+	val = e1000_mii_readreg(dev, ep->mii_if.phy_id, 0x14);
+	log_boot("PHY reg # 0x14  after reset : 0x%x\n", val);
+	val = e1000_mii_readreg(dev, ep->mii_if.phy_id, 0x15);
+	log_boot("PHY reg # 0x15  after reset : 0x%x\n", val);
+
+	val = e1000_mii_readreg(dev, ep->mii_if.phy_id, PHY_BIST_CFG2);
+	log_boot("PHY reg # 0x1a (BIST_CFG2): 0x%x\n", val);
+	val |= LINK_SEL;
+	e1000_mii_writereg(dev, ep->mii_if.phy_id, PHY_BIST_CFG2, val);
+	val = e1000_mii_readreg(dev, ep->mii_if.phy_id, PHY_BIST_CFG2);
+	log_boot(" read mgio csr #0x04 (BIST_CFG2) : 0x%x\n", val);
+	/* move e1000 link status select to default 0 link */
+	val = e1000_read_mgio_csr(ep);
+	val &= ~LSTS;
+	val |= SLSP;
+	e1000_write_mgio_csr(ep, val);
+	log_boot(" write mgio csr #0x04 (BIST_CFG2): 0x%x\n", val);
+
+	return 0;
+}
+
+static int l_e1000_open(struct net_device *dev) {
+	log_debug("mac addr: %x:%x:%x:%x:%x:%x\n", dev->dev_addr[0], dev->dev_addr[1], dev->dev_addr[2], dev->dev_addr[3], dev->dev_addr[4], dev->dev_addr[5]);
+	l_e1000_reset(dev);
+	return 0;
+}
+
+static int l_e1000_xmit(struct net_device *dev, struct sk_buff *skb) {
+	struct l_e1000_priv *ep = netdev_priv(dev);
+	uint16_t status;
+	int len;
+
+
+	if (skb->len < 0x40) {
+		len = 0x40;
+	} else {
+		len = skb->len;
+	}
```

Comment:
What is it?

Reply:
It's minimal ethernet packet length. It's not correct to use frames smaller size.

Reply:
Hm, and what will be if we extend skb len to 0x40 in this function? Didn't it will send some extra 'trash' in the packet?

Reply:
In other words, as I can see this function can accept `skb->len < 0x40`, because there is no assertion here. But since this function can accept `skb->len < 0x40` is it correct to pass greater length to tx descriptor?

Reply:
> Didn't it will send some extra 'trash' in the packet

Yes, it does. But it doesn't matter. ARP has 42 length, but  In this net card DMA does not start if frame len < 0x40,  it saves for this problem.